### DISCUSSION TITLE
Very bizarre behavior - environment variables disappearing

### DISCUSSION THREAD

**Discussion by thepuzzlemaster** [dokku/dokku#6882] [Q&A]
Okay, about to take you on a small journey here. Looking to see if anyone has any insight into what might have possibly happened to us.

As outlined in this issue https://github.com/dokku/dokku/discussions/6726, our server slowly fills up as we push new code to it.

Our server setup includes 3 apps (2 python and 1 nextjs), with 1 of the python apps linked to a Postgres db, which is also housed on the same server. 
On April 23rd we were pushing to both the python app and the nextjs app at the same time. While we were pushing to both apps, our drive filled with space and both apps failed to push.

We ran `docker system prune -a` to clear up the extra space, which is what we usually do. However when we were done with the prune, one of the python apps was cleared out. Almost all our config values were wiped. There were about 4 remaining, but I think they were just the default Heroku variables. Our 2nd python app was completely gone. The entire app didn't even exist anymore.

We were never able to track down any logs which would have explained what happened. Nor were we ever able to identify what might have caused it in any other way. We host on digital ocean, and it's possible there was some sort of issue with our droplet, or the storage volume associated with that droplet. But we didn't see anything on the digital ocean side which explained it either.

Anyways, it's definitely a longshot, but curious if anyone has ever experienced anything like this in their dokku apps before. Or have any thoughts as to what might have happened. Or a way to diagnose what might have happened. We tried reproducing, but weren't able to.

**Comment by josegonzalez** [comment]
Apps are stored in the `/home/dokku` folder (with app-config/data split between `/var/lib/dokku/config` and `/var/lib/dokku/data`). These aren't touched by `docker` commands, so maybe you have some other command that you ran which deleted large folders or something, and inadvertenly deleted those apps?

Note: if you think they were dokku commands, you could enable the `events` plugin. That logs _all_ internal plugin trigger calls, which could help indicate what commands were run that deleted the apps (since there are triggers related to app deletion).

**Comment by thepuzzlemaster** [comment]
Super old thread here, but we had done a bit more research into the docker prune commands as a result of still dealing with this issue https://github.com/dokku/dokku/discussions/6726. Admittedly, we are still running v0.33.3, so it's possible an update might help things there.

But regardless, it seems like what might have happened here, is that when we pushed our application and it ran out of space, the container got stopped for some reason. Then when we ran `docker system prune -a` it cleared out the stopped application entirely, which would match what we saw on our server - that the application had pretty much just entirely disappeared.

With a more clear understanding of the `-a` option, we will avoid that in the future. It also turns out that a significant portion of the space usage is build cache, not images or containers, so we will be using `docker builder prune` instead, which should be safer as well.

Just wanted to provide an update in case anyone else comes across something similar.

**Reply by josegonzalez** [reply]
Environment variables are stored in the `/home/dokku/$APP/ENV` file. This is due to be moved to the properties system at some point, at which point it will live somewhere in `/var/lib/dokku/config`.

A `docker system prune` does not touch the Dokku home folder, regardless of whatever command you are running. That also goes for the `/var/lib/dokku` folder.

I would hazard a guess that you actually deleted the app folder (not knowing what it was for) while trying to clean up apps. This caused the app to "disappear", resulting in a loss of environment variables.

**Reply by thepuzzlemaster** [reply]
Ahhh, that makes sense. Yeah, it wouldn't make sense for environment variables to be stored in the containers now that I'm thinking about it. Thanks for chiming in with your ever valuable insight!

**Comment by bakatz** [comment]
@josegonzalez Chiming in here, I seem to have hit something very similar last night: https://github.com/dokku/dokku/discussions/6726#discussioncomment-10667710
